fairseq/training.py

```diff
@@ -0,0 +1,502 @@
+import logging
+import os
+import subprocess as sp
+import yaml
+import sys
+import copy
+import collections.abc
+import re
+from typing import Dict, Any
+
+from sisyphus import Task, Job, tk
+from sisyphus.hash import sis_hash_helper
+
+import i6_core.util as util
+
+
+class FairseqHydraConfig:
+    """
+    An object that manages a Fairseq hydra config.
+    """
+
+    def __init__(
+        self,
+        config_dict: Dict[str, Any],
+        post_config_dict: Optional[Dict[str, Any]] = None,
+        *,
+        package_name: str = "",
+    ):
+        """
+        :param config_dict: Contains the information which is needed for fairseq-hydra-train. Will be converted and dumped into a .yaml
+        :param dict post_config_dict: dictionary of the FairseqHydraConfig config variables that are not hashed
+        :param package_name: The @package directory that is required to be added to the top of Hydra config, for example "# @package _group_"
+        """
+        assert isinstance(config_dict, dict)
+        assert isinstance(post_config_dict, dict) or post_config_dict is None
+        self.config_dict = config_dict
+        self.post_config_dict = post_config_dict if post_config_dict is not None else {}
+        self.package_name = package_name
+        self.check_consistency()
+
+    @staticmethod
+    def update_nested_dict(dict1, dict2):
```

Comment:
Should we make this a general helper function? I don't think it's really related to the `FairseqHydraConfig`.

Reply:
Yes, it is not very related to FairseqHydraConfig, so I use the static method here. But we could make this a general helper function, where would you suggest to put the general helper function to?

Reply:
I don't have any strong feelings here. You could e.g. just make it a function outside the class.

Reply:
I would propose `util.py`

Reply:
but actually.. there must already be an existing implementation in the internet.. I would guess we are not the first to want to implement nested dict update..

Reply:
>  there must already be an existing implementation in the internet

Haha, I did a little research, found a nice implementation and noticed that you indeed found and used the same. Good job, nevermind my previous comment xD

Reply:
ah yes, I have searched in stack flow and found this implementation. I have put this function in util now xD